### PR TITLE
Fixes lazy loading for .NET

### LINKED ISSUE #2834: No metrics for `HttpClient` are created when tracing is enabled
# Bug Report

## Symptom

**Describe the bug**
When using version v1.0.0-rc.2 and both metrics and traces are enabled, no metrics from the `OpenTelemetry.Instrumentation.Http` instrumentation library are created. Traces are created.

When traces are disabled (`OTEL_DOTNET_AUTO_TRACES_ENABLED=false`) HTTP metrics are created. 

However, it seems there's no way to export both metrics and traces.

**Runtime environment (please complete the following information):**

- OpenTelemetry Automatic Instrumentation version: 1.0.0-rc.2
- OS: Ubuntu 22
- .NET version: .NET 7

**Additional context**
After some debugging, I think that's what's happening:
* HTTP metrics instrumentation is [supposed to be initialized when `System.Net.Http` is loaded](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/src/OpenTelemetry.AutoInstrumentation/Loading/Initializers/HttpClientMetricsInitializer.cs#L25).
* The `HttpClientMetricsInitializer` mentioned above is registered during a call to [MeterProviderBuilder.UseEnvironmentVariables](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs#L142).
* However, `System.Net.Http` is already loaded before when the [traces pipeline is built](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/main/src/OpenTelemetry.AutoInstrumentation/Instrumentation.cs#L123), likely by initializing exporters for tracing or HTTP instrumentation for tracing.

TLDR; At the moment when `HttpClientMetricsInitializer` registers its event handler that should fire when the `System.Net.Http` assembly is loaded, the assembly is already loaded.

## Reproduce

 1. Set environment variables to use .NET auto-instrumentation, enable both traces and metrics with the default OTLP exporter.
 2. Also export to the console:
   ```
   OTEL_DOTNET_AUTO_TRACES_CONSOLE_EXPORTER_ENABLED=true  
   OTEL_DOTNET_AUTO_METRICS_CONSOLE_EXPORTER_ENABLED=true
   ```
 3. Run an example application that makes an HTTP request using `HttpClient`.
 4. No HTTP metrics are exported.

### PR DESCRIPTION
**Pull request by RassK**
## Why

Fixes #2834

## What

* Enables checks against race conditions / multiple lazy loads also in .NET (previously done for .NET Fx) 
* Small optimization for search code

## Tests

Existing.

## Checklist

- [x] `CHANGELOG.md` is updated.
- [ ] Documentation is updated.
- [x] New features are covered by tests.

### CODE REVIEW

**Comment by Kielek** [comment]
Please add changelog entry with bugfix description.

**Review by pyohannes — APPROVED** [review]
Tested, I confirm it fixes the issue initially reported.

**Review by Kielek — APPROVED** [review]

**Review by nrcventura — APPROVED** [review]

**Comment by pjanotti** [comment]
@RassK Is this covered by some test?

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs [hunk 1]
```diff
@@ -59,23 +56,23 @@ public OnAssemblyLoadInitializer(ILifespanManager lifespanManager, Instrumentati
...
             if (isRequiredAssemblyLoaded)
             {
                 OnRequiredAssemblyDetected();
             }
-#endif
+        }
+
+        private static string? GetAssemblyName(Assembly assembly)
+        {
+            return assembly.FullName?.Split(new[] { ',' }, count: 2)[0];
```
[the remark this answers is not in the document]
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs]: with spans this shows very good results 👍

**Comment by RassK** [comment]
> @RassK Is this covered by some test?

@pjanotti came later to this idea > https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/2835/commits/be38e8ae46a4ebd915ff28289207db184b85b66b

**Review by pjanotti — APPROVED** [review]

**Review by pjanotti — APPROVED** [review]

**Pull request merged**